### PR TITLE
Feature/shared resources

### PR DESCRIPTION
**Pull request by weiyuan-jiang**
Move shared resource to LoggerManager and free resources through LoggerManager instead of through individual logger

### CODE REVIEW

File: src/Logger.F90 [hunk 1]
```diff
@@ -155,16 +155,16 @@ end function get_name
 !---------------------------------------------------------------------------
    subroutine add_handler(this, handler)
       class (Logger), intent(inout) :: this
-      class (AbstractHandler), intent(in) :: handler
+      class (AbstractHandler), pointer, intent(in) :: handler
```
Comment (tclune) [inline src/Logger.F90:158]: This just needs to be ```TARGET```, not ```POINTER```.

File: src/Logger.F90 [hunk 2]
```diff
@@ -195,18 +195,26 @@ end subroutine free
 !---------------------------------------------------------------------------
    subroutine remove_handler(this, handler)
       class (Logger), intent(inout) :: this
-      class (AbstractHandler), intent(in) :: handler
+      class (AbstractHandler), pointer, intent(in) :: handler
```
Comment (tclune) [inline src/Logger.F90:198]: ```TARGET``` (as above)

File: src/LoggerManager.F90 [hunk 3]
```diff
@@ -343,36 +346,33 @@ subroutine load_config(this, cfg, unused, extra, comm)
...
+      if (.not. subcfg%is_none()) call this%elements%build_formatters(subcfg, extra=extra)
 
       call cfg%get(subcfg, 'handlers')
-      if (.not. subcfg%is_none()) call elements%build_handlers(subcfg, extra=extra)
+      if (.not. subcfg%is_none()) call this%elements%build_handlers(subcfg, extra=extra)
 
-      call this%build_loggers(cfg, elements, extra=extra)
-      call this%build_root_logger(cfg, elements, extra=extra)
+      call this%build_loggers(cfg, extra=extra)
+      call this%build_root_logger(cfg, extra=extra)
```
Comment (tclune) [inline src/LoggerManager.F90:368]: Use an ASSOCIATE block:
```f90
associate (elements => this%elements)
...
end associate
```
This allows the code to remain mostly unchanged.

**Comment by tclune** [comment]
The other change to this PR that somehow did not make it into the review:

We need two types of Handler containers.  The first uses ALLOCATABLE (as per the original code), and the other uses POINTER (as per how this PR has done it).    LoggerManager should have  a HandlerVector, and Logger's should have HandlerPtrVector's.

**Comment by weiyuan-jiang** [comment]
LoggerManager has a handleMap which uses allocatable. So We don't need to worry about it.

**Comment by weiyuan-jiang** [comment]
@tclune I have made some minor changes and tested it with the logging.yaml posted by Ben. It works.

File: src/FileHandler.F90 [hunk 4]
```diff
@@ -255,7 +255,7 @@ logical function is_lockable(this)
       class (FileHandler), intent(in) :: this
 
       is_lockable = .false.
-      if (associated(this%lock)) is_lockable = this%lock%is_initialized()
+      if (allocated(this%lock)) is_lockable = this%lock%is_initialized()
```
Comment (tclune) [inline src/FileHandler.F90:258]: Why is the check on initialized() necessary?   It was not before.

Comment (weiyuan-jiang) [reply src/FileHandler.F90:258]: You can allocate a lock without initialization. So it is wise to check. I don't know when you switch to this lazy initialization.

Comment (tclune) [reply src/FileHandler.F90:258]: ok - I'll provisionally accept it then.

**Pull request merged**